### ISSUE TITLE
Exception while deserializing nested Map<String,Object> objects

### ISSUE THREAD

**Issue by christian-lechner** [arangodb/spring-data#26]
Hello!

I found an error while deserializing nested maps. See the following class:
```java
public class Test {
    @Id
    private String id;

    @Key
    private String key;

    private final Map<String, Object> attrs = new HashMap<>();

    public put(String key, Object value) {
        attrs.put(key, value);
    }

    // getter & setter skipped
}
```
```java
Test test = new Test();
Map<String, Object> innerMap = new HashMap<>();
innerMap.put("key", "value");
test.put("map", innerMap);
arangoOperations.insert(test);
arangoOperations.find(test.getKey(), Test.class).get();
```

This raises a `org.springframework.data.mapping.MappingException: No mapping metadata found for type java.lang.Object` exception at `DefaultArangoConverter.java:168`.

The exception has its root in the following method of `DefaultArangoConverter`:
```java
private Object read(final TypeInformation<?> type, final DBEntity source) {
	// ...
	final Optional<? extends ArangoPersistentEntity<?>> entity = Optional
			.ofNullable(context.getPersistentEntity(type.getType()));
	return read(type, source, entity);
}
```
As a side note: `Optional` was not intended to be used as arguments (see [here](http://blog.joda.org/2014/11/optional-in-java-se-8.html)). Also it would be more clear to raise the exception where it actually happened (at `context.getPersistentEntity()`). There is also a nice method from Spring Data's `MappingContext`, which raises an exception for you: `context.getRequiredPersistentEntity()`.

To fix the problem with nested maps and lists, we could check if the type is `java.lang.Object` and create a map if the source is of type `DBDocumentEntity` or a list if the source is of type `DBCollectionEntity`.

What do you think?

**Comment by christian-lechner** [comment]
As `DBDocumentEntity` inherits from `HashMap` and `DBCollectionEntity` inherits from `ArrayList`, one could simply do the following in `Object read(final TypeInformation<?> type, final DBEntity source)`:
```java
if (type.equals(ClassTypeInformation.OBJECT)) {
	return source;
}
```
Then the next issue that arises is: what if the stored object was not of type `Map`? To fix this we need to add type hints to the object in the db. But this is already discussed in #17 .

**Comment by christian-lechner** [comment]
This issue would also be fixed by #33.

**Comment by christian-lechner** [comment]
Fixed by #33.
